### ISSUE TITLE
Questions on online quantization

### ISSUE THREAD

**Issue by lzhangzz** [spcl/QuaRot#11]
Nice paper and repo! 

I'm trying to get a clear understanding of QuaRot to see if we can integrate it into [LMDeploy](https://github.com/InternLM/lmdeploy). I have some questions on the online quantization part

1. It seems that all hadamards for online quantization are not randomized. If I'm understanding it right, the randomness brought by the global randomized hadamard transformation `Q` has no effective on the FP16 components inside a FFN/attention block. Is the quality of online quantization not impacted by the randomness or we can get more performance by using full RHT here?

2. Can we omit the partial hadamard before `o_proj` by applying only head-wise hadamard rotation to values like how queries and keys are processed? I'm confused on why it must be done to ensure correctness.

3. Is the second `d_h` in the first line a typo?
![20240426-203544](https://github.com/spcl/QuaRot/assets/19465119/33d4048a-1107-4a35-8cb8-d1c81c4ab3b8)

Thanks very much!

**Comment by sashkboos** [comment]
@lzhangzz 

Thank you so much for your questions. 


>It seems that all hadamards for online quantization are not randomized. If I'm understanding it right, the randomness brought by the global randomized hadamard transformation Q has no effective on the FP16 components inside a FFN/attention block. Is the quality of online quantization not impacted by the randomness or we can get more performance by using full RHT here?

We have done a few experiments to understand this. It seems that you can get better accuracy by using RHT. However, we have a fast CUDA kernel only for Hadamard transformations, and using RHT results in slower transformations. So we used RHT when we want to fuse the transformations into the weights (so this will have no overhead during the inference) and deterministic hadamards for our online transformations.

>Can we omit the partial hadamard before o_proj by applying only head-wise hadamard rotation like how queries and keys are processed? I'm confused on why it must be done to ensure correctness.

We fuse the inverse of the (full) Hadamard in o_proj weights. This means that the input of o_proj should be transformed using a full Hadamard transformation. To this end, we fuse per-head transformations into the v_proj weights so we need to apply outer-head transformation online to get full Hadamard transformation. 

>Is the second d_h in the first line a typo?

No. This is right. I just explained it in the previous question. 

Please let me know if you have any other question :)

**Comment by sashkboos** [comment]
@lzhangzz 

Thanks

> So have you tried to use only the head-wise transformation for the values and see how it impact performance?

Actually, this is a very good question. We started with this but the accuracy of per-head rotation was bad! However, another option is to rotate per-head and then quantize the inputs using group-wise quantization (instead of row-wise in the current setting) with group-wise=head_dim. This can solve the issue (we have done some experiments and I can verify it) in the cost of doing group-wise quantization in o_proj. 

In conclusion, it is important to apply quantization on the same granularity as Hadamard. If you only apply Hadamards on each head separately, the scales between different heads will be different and quantizing across the whole heads leads to poor accuracy.

**Comment by lzhangzz** [comment]
> However, another option is to rotate per-head and then quantize the inputs using group-wise quantization (instead of row-wise in the current setting) with group-wise=head_dim.

This is great. It's useful in tensor parallel mode where you don't have access to all channels without communicating with other ranks.
